[PWA] Installation guide (#25058)
mdn/content: files/en-us/web/progressive_web_apps/guides/installation/index.md

---

> title: Installation

Author: Not sure I love this title but I couldn't think of anything better :).

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <link rel="manifest" href="manifest.json" />
    <!-- ... -->
  </head>
  <body></body>
</html>
```

If the PWA has more than one page, every page must reference the manifest in this way.

The manifest contains a single JSON object containing a collection of members, each of which defines some aspect of the PWA's appearance or behavior. Here's a rather minimal manifest, containing just two members: `"name"` and `"icons"`.

```json
{
  "name": "My PWA",
  "icons": [
    {
      "src": "icons/512.png",
      "type": "image/png",
      "sizes": "512x512"
    }
  ]
}
```

Chromium-based browsers, including Google Chrome, Samsung Internet, and Microsoft Edge, require that the manifest includes the following members:

- [`name`](/en-US/docs/Web/Manifest/name)
- [`icons`](/en-US/docs/Web/Manifest/icons)
- [`start_url`](/en-US/docs/Web/Manifest/start_url)
- [`display`](/en-US/docs/Web/Manifest/display) and/or [`display_override`](/en-US/docs/Web/Manifest/display_override)

For a full description of every member, see the [web app manifest reference documentation](/en-US/docs/Web/Manifest).

### HTTPS

> For a web app to be installable, it must be served over HTTPS.

Reviewer: It might be worth mentioning something like: "For local development purposes, browsers do not require HTTPS if the web app is accessed via the `localhost` domain".

On mobile:

- Firefox, Chrome, Edge, Opera, and Samsung Internet Browser all support installing PWAs on Android.
- Only Safari is allowed to install PWAs on iOS.

### Triggering the install prompt

A PWA can provide its own UI for the user to open the install prompt, instead of relying on the UI provided by the browser. This enables a PWA to provide some context and a reason for the user to install the PWA.

This technique relies on the [`beforeinstallprompt`](/en-US/docs/Web/API/Window/beforeinstallprompt_event) event, which is fired on the global [`Window`](/en-US/docs/Web/API/Window) object as soon as the browser has determined that the PWA is installable. This event has a [`prompt()`](/en-US/docs/Web/API/BeforeInstallPromptEvent/prompt) method that shows the install prompt. So a PWA can:

- add its own "Install" button
- listen for the `beforeinstallprompt` event
- cancel the event's default behavior by calling [`preventDefault()`](/en-US/docs/Web/API/Event/preventDefault)
- in the event handler for its own "Install" button, call [`prompt()`](/en-US/docs/Web/API/BeforeInstallPromptEvent/prompt).

> ### Customizing installation

Reviewer: ```suggestion
### Customizing the installation prompt
```

By default, the install prompt contains the name and icon for the PWA. If you provide values for the [`description`](/en-US/docs/Web/Manifest/description) and [`screenshots`](/en-US/docs/Web/Manifest/screenshots) manifest members, then, on Android only, these values will be shown in the install prompt, giving the user extra context and motivation to install the PWA.

> The screenshot below shows what the install prompt for the [pwamp demo](https://github.com/MicrosoftEdge/Demos/tree/main/pwamp) looks like on Google Chrome, running on Android:

Reviewer: ```suggestion
The screenshot below shows what the install prompt for the [PWAmp demo](https://github.com/MicrosoftEdge/Demos/tree/main/pwamp) looks like on Google Chrome, running on Android:
```

Really minor, but I'm quite attached to the spelling of PWAmp, wich is meant to be read as P-W-Amp (sort of like WinAmp, but PW instead of Win :) )

That being said, once we have a coupld of demo PWAs (for the tutorial section), we might want to use them here instead.

![Install prompt for pwamp on Android](pwamp-install-prompt-android.png)

## Launching the app

> Once the PWA is installed its icon is shown on the device alongside any other apps that the user has installed, and clicking the icon launches the app.

Reviewer: ```suggestion
Once the PWA is installed, its icon is shown on the device alongside any other apps that the user has installed, and clicking the icon launches the app.
```

---
title: Installation
slug: Web/Progressive_web_apps/Guides/Installation
---

> One of the defining aspects of a PWA is that it can be installed on the device, and then appears to users as a platform-specific app, a permanent feature of their device which they can launch directly from the platform like any other app.

Reviewer: ```suggestion
One of the defining aspects of a PWA is that it can be installed on the device, and then appears to users as a platform-specific app, a permanent feature of their device which they can launch directly from the operating system like any other app.
```

I find the word platform here a bit hard to understand. OS seems more intuitive.

We can summarize this as follows:

- The PWA can be installed from the platform's app store or installed directly from the web.
- The PWA can be installed like a platform-specific app, and can customize the install process.
- Once installed, the PWA gets an app icon on the device, alongside platform-specific apps.
- Once installed, the PWA can be launched as a standalone app, rather than a website in a browser.

We'll discuss each of these features in this guide. First, though, we'll discuss the requirements that a website must meet for it to be installable.

## Installability

> For a website to be installable it needs to meet some technical requirements. We can consider these the minimum requirements for a website to be a PWA.

Reviewer: Since we're talking about apps here, the word website seems a bit out of place for me and I think I'd prefer web app.

```suggestion
For a web app to be installable, it needs to meet some technical requirements. We can consider these the minimum requirements for a web app to be a PWA.
```

### The web app manifest

A web app manifest is a JSON file that tells the browser how the PWA should appear and behave on the device. For a website to be a PWA it must be installable, and for it to be installable it must include a manifest.

The manifest is included using a {{HTMLElement("link")}} element in the app's HTML:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <link rel="manifest" href="manifest.json" />
    <!-- ... -->
  </head>
  <body></body>
</html>
```

If the PWA has more than one page, every page must reference the manifest in this way.

> The manifest contains a single JSON object containing a collection of keys, each of which defines some aspect of the PWA's appearance or behavior. Here's a rather minimal manifest, containing just two keys: `"name"` and `"icons"`.

Reviewer: In manifest language, we call the top-level properties `members`, not `keys`. Using members here would make things consistent with specs and other docs out there.

Author: Ha, looking at your how-to last week I was going to complain that you use "members" instead of "keys" but then I saw that was what the spec (and most of the MDN manifest reference) uses. So yes.

```json
{
  "name": "My PWA",
  "icons": [
    {
      "src": "icons/512.png",
      "type": "image/png",
      "sizes": "512x512"
    }
  ]
}
```

For a full description of every key, see the [web app manifest reference documentation](/en-US/docs/Web/Manifest).

> ### Additional installability requirements

Reviewer: HTTPS and service worker are both, for now, required by all browsers to consider an app as installable. Together with manifest, these 3 criteria have been part of the PWA's DNA from the start, and I think they should each have an h3-level section like the one you used for manifest.

The UI for installing a PWA from the web varies from one browser to another, and from one platform to another. For example, a browser might include an "Install" icon in the URL bar when the user navigates to the page:

![Chrome URL bar, showing PWA install icon](pwa-install.png)

When the user selects the icon, the browser displays a prompt asking if they want to install the PWA, and if they accept, the PWA is installed.

The prompt displays the name and icon for the PWA, taken from the [`name`](/en-US/docs/Web/Manifest/name) and [`icons`](/en-US/docs/Web/Manifest/icons) manifest keys.

### Browser support

Support for installing PWAs from the web varies by browser and by platform.

On desktop:

- Firefox and Safari do not support installing PWAs on any desktop operating systems.
- Chrome and Edge support installing PWAs on Linux, Windows, macOS, and Chromebooks.

On mobile:

- Firefox, Chrome, Edge, Opera, and Samsung Internet Browser all support installing PWAs on Android.

> - Only Safari is allowed to install PWAs on iOS.

Reviewer: Although this will change with the latest version of webkit. See https://webkit.org/blog/13878/web-push-for-web-apps-on-ios-and-ipados/ and search for "Third-party browser support for Add to Home Screen".

This technique relies on the [`beforeinstallprompt`](/en-US/docs/Web/API/Window/beforeinstallprompt_event) event, which is fired on the global [`Window`](/en-US/docs/Web/API/Window) object as soon as the browser has determined that the PWA is installable. A PWA can:

- add a UI element for the user to install the PWA (for example, its own "Install" button)
- listen for the `beforeinstallprompt` event
- cancel the event's default behavior by calling [`preventDefault()`](/en-US/docs/Web/API/Event/preventDefault)
- keep a reference to the [`BeforeInstallPromptEvent`](/en-US/docs/Web/API/BeforeInstallPromptEvent), and in the event handler for its own "Install" button, call the [`BeforeInstallPromptEvent.prompt()`](/en-US/docs/Web/API/BeforeInstallPromptEvent/prompt) method. This method will show the browser prompt that asks the user if they want to install the PWA.

For example:

```js
window.addEventListener("beforeinstallprompt", (event) => {
  // Don't let the default prompt go.
  event.preventDefault();

  // Instead, wait for the user to click the install button.
  installButton.addEventListener("click", () => event.prompt());
});
```

> The `prompt()` method returns a {{jsxref("Promise")}} that resolves with a string `"accepted"` or `"dismissed"` indicating whether or not the user chose to install the PWA.

Author: Perhaps this is too instructional and belongs in a How-to?

Reviewer: Yes we have the custom install flow as part of the How-To section in our shared doc. Maybe it can be here too, only shorter? In any case, we should link to the How-To here, once it exists.

Author: Right, I'll keep something here but remove the sample code. Later on we can add links to the How-to pages for the details.

> ### Customising installation

Reviewer 2: ```suggestion
### Customizing installation
```

> By default, the install prompt contains the name and icon for the PWA. If you provide values for the [`description`](/en-US/docs/Web/Manifest/description) and [`screenshots`](/en-US/docs/Web/Manifest/screenshots) manifest keys, then, on Android only, these values will be shown in the install prompt, giving the user extra context and motivation to install the PWA.

Reviewer: ```suggestion
By default, the install prompt contains the name and icon for the PWA. If you provide values for the [`description`](/en-US/docs/Web/Manifest/description) and [`screenshots`](/en-US/docs/Web/Manifest/screenshots) manifest members, then, on Android only, these values will be shown in the install prompt, giving the user extra context and motivation to install the PWA.
```

(probably worth doing a search/replace on the entire document for keys->members).

Also, I think Chrome for desktop now also uses screenshots and description. I definitely saw it a few months ago, but now it seems disabled. So I think they're working on it behind a flag.

Author: Yes, when I tested last week it didn't seem to be supported in Chrome desktop. I think compat is challenging in this project, and we should have a separate compat page where we document all such issues.

The screenshot below shows what the install prompt for the [pwamp demo](https://github.com/MicrosoftEdge/Demos/tree/main/pwamp) looks like on Google Chrome, running on Android:

![Install prompt for pwamp on Android](pwamp-install-prompt-android.png)

## Launching the app

Once the PWA is installed its icon is shown on the device alongside any other apps that the user has installed, and clicking the icon launches the app.

You can use the [`display`](/en-US/docs/Web/Manifest/display) manifest key to control the _display mode_: that is, how the PWA appears when it is launched. In particular:

- `"standalone"` indicates that the PWA should look and feel like a platform-specific application, with no browser UI elements
- `"browser"` indicates that the PWA should be opened as a new browser tab or window, just like a normal website.

> If the browser does not support a given display mode, `display` will fall back to a supported display mode according to a predefined sequence. The [`display_override`](/en-US/docs/Web/Manifest/display_override) enables you to redefine the fallback sequence.

Author: This seems like an OK level of detail, and we could link to the How-to for specific instructions and example code?

Reviewer: Yes, I think that's a good idea. Let's add the link when we can with a leading sentence like "For more information about the various display modes and how they fallback, see ..."
